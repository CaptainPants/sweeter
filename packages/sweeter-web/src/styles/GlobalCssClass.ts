import { type StylesheetDependencyProvider } from './StylesheetDependencyProvider.js';
import { preprocessClassContent } from './preprocessor/preprocess.js';
import type {
    GlobalStyleSheetContentGeneratorContext,
    AbstractGlobalCssStylesheet,
    StylesheetGenerator,
    AbstractGlobalCssClass,
} from './types.js';

type GlobalCssClassContent =
    | ((self: GlobalCssClass) => string | StylesheetGenerator)
    | string;

type ContentConstructed = StylesheetGenerator | string;

export interface GlobalCssClassOptions {
    /**
     * Base the class name for this class on this value (it will be prefixed/suffixed or otherwise made unique).
     */
    className: string;
    content?: GlobalCssClassContent | undefined;
    preprocess?: boolean;
    extraDependencies?: StylesheetDependencyProvider;
}

export class GlobalCssClass
    implements AbstractGlobalCssStylesheet, AbstractGlobalCssClass
{
    public readonly className: string;
    public readonly id: string;
    public readonly symbol: symbol;
    public readonly classContent?: ContentConstructed | undefined;
    public readonly preprocess: boolean;

    #extraDependencies?: AbstractGlobalCssStylesheet[];

    constructor(options: GlobalCssClassOptions);
    constructor({
        className,
        content,
        preprocess = true,
        extraDependencies,
    }: GlobalCssClassOptions) {
        this.className = className;
        this.id = className;
        this.symbol = Symbol('GlobalCssClass-' + className);
        this.preprocess = preprocess;

        this.classContent =
            typeof content === 'function' ? content(this) : content;

        if (extraDependencies) {
            extraDependencies.addDependencyListener((dependency) => {
                this.#extraDependencies ??= [];
                this.#extraDependencies.push(dependency);
            });
        }
    }

    getContent(
        context: GlobalStyleSheetContentGeneratorContext,
    ): string | undefined {
        const name = context.getPrefixedClassName(this);

        const content =
            typeof this.classContent === 'string'
                ? this.classContent
                : this.classContent?.(context);

        if (!content) {
            return undefined;
        }

        const transformed = this.preprocess
            ? preprocessClassContent(name, content)
            : `.${name}\r\n{\r\n${content}\r\n}`;
        return transformed;
    }

    getReferencedStylesheets():
        | readonly AbstractGlobalCssStylesheet[]
        | undefined {
        let result: AbstractGlobalCssStylesheet[] | undefined =
            typeof this.classContent === 'function'
                ? [...this.classContent.references]
                : undefined;

        if (this.#extraDependencies) {
            result ??= [];
            result.push(...this.#extraDependencies);
        }
        return result;
    }
}
