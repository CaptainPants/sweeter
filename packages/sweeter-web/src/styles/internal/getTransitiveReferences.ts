import { type AbstractGlobalCssStylesheet } from '../types.js';

export function getTransitiveReferences(
    stylesheet: AbstractGlobalCssStylesheet,
) {
    const sheets: AbstractGlobalCssStylesheet[] = [stylesheet];

    const referenced = new Set([stylesheet]);

    // Recursively add referenced sheets
    for (let i = 0; i < sheets.length; ++i) {
        const current = sheets[i]!;

        const references = current.getReferencedStylesheets();
        if (references) {
            for (const currentReference of references) {
                if (!referenced.has(currentReference)) {
                    sheets.push(currentReference);
                    referenced.add(currentReference);
                }
            }
        }
    }

    return sheets;
}
