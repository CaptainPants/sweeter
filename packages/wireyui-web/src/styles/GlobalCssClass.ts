import type { GlobalStyleSheet } from './types.js';

let counter = 0;

export class GlobalCssClass implements GlobalStyleSheet {
    public readonly className: string;
    public readonly content: string;
    public readonly id: symbol;

    constructor(options: { id: string; content: string }) {
        this.className = '_glbl' + counter + '_' + options.id;
        ++counter;

        this.content = `.${this.className} {\r\n${options.content}\r\n}`;
        this.id = Symbol('GlobalCssClass');
    }
}
