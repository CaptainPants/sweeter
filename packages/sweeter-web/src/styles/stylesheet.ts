import { type StylesheetGenerator } from './types.js';
import { type GlobalCssClass } from './GlobalCssClass.js';

/**
 * Template string function.
 *
 * Creates a StylesheetGenerator based on a templated string, where you can use GlobalCssClass references in style selectors.
 * @param template
 * @param params
 * @returns
 */
export function stylesheet(
    template: TemplateStringsArray,
    ...params: (GlobalCssClass | StylesheetGenerator | string)[]
): StylesheetGenerator {
    return (context) => {
        const res: string[] = [];
        const last = template.length - 1;
        for (let i = 0; i < template.length; ++i) {
            res.push(template[i]!);

            if (i !== last) {
                const param = params[i]!;

                if (typeof param === 'string') {
                    res.push(param);
                } else if (typeof param === 'function') {
                    res.push(param(context));
                } else {
                    res.push('.');
                    res.push(context.getPrefixedClassName(param));
                }
            }
        }
        return res.join('');
    };
}
