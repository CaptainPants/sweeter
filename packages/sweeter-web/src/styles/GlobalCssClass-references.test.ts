import { GlobalCssClass } from './GlobalCssClass.js';
import { getTransitiveReferences } from './internal/getTransitiveReferences.js';
import { stylesheet } from './stylesheet.js';

it('Self reference', () => {
    const classA = new GlobalCssClass({
        className: 'test',
        content: (self) => {
            return stylesheet`
                color: green;

                .root .${self} {
                    color: red;
                }

                & .root & {
                    color: blue;
                }

                & .root & .${self} {
                    color: orange;
                }

                .inner {
                    .${self} {
                        color: rebeccapurple;
                    }
                }
            `;
        },
    });

    const content = classA.getContent({
        getPrefixedClassName: (class_) => class_.className,
    });

    expect(content).toMatchSnapshot();
});

it('Reference to another class', () => {
    const classA = new GlobalCssClass({
        className: 'a',
        content: (self) => {
            return stylesheet`
                color: red;
            `;
        },
    });

    const classB = new GlobalCssClass({
        className: 'b',
        content: () => {
            return stylesheet`
                color: green;

                .${classA} {
                    color: blue;
                }
            `;
        },
    });

    const content = classB.getContent({
        getPrefixedClassName: (class_) => class_.className,
    });

    expect(content).toMatchSnapshot();

    expect(getTransitiveReferences(classB)).toStrictEqual([classB, classA]);
});

it('Transitive reference to a 3rd class', () => {
    const classA = new GlobalCssClass({
        className: 'a',
        content: (self) => {
            return stylesheet`
                color: red;
            `;
        },
    });

    const classB = new GlobalCssClass({
        className: 'b',
        content: () => {
            return stylesheet`
                color: green;

                .${classA} {
                    color: blue;
                }
            `;
        },
    });

    const classC = new GlobalCssClass({
        className: 'b',
        content: () => {
            return stylesheet`
                color: yellow;

                .${classB} {
                    color: rebeccapurple;
                }
            `;
        },
    });

    const content = classC.getContent({
        getPrefixedClassName: (class_) => class_.className,
    });

    expect(content).toMatchSnapshot();

    expect(getTransitiveReferences(classC)).toStrictEqual([
        classC,
        classB,
        classA,
    ]);
});
