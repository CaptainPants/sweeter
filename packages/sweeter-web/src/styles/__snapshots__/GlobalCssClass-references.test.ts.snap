// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reference to another class 1`] = `
".b {
    color: green;
}
.b .a {
    color: blue;
}
"
`;

exports[`Self reference 1`] = `
".test {
    color: green;
}
.test .root .test {
    color: red;
}
.test .root .test {
    color: blue;
}
.test .root .test .test {
    color: orange;
}
.test .inner .test {
    color: rebeccapurple;
}
"
`;

exports[`Transitive reference to a 3rd class 1`] = `
".b {
    color: yellow;
}
.b .b {
    color: rebeccapurple;
}
"
`;
